### PR TITLE
Full Project code review

### PR DESCRIPTION
**Pull request by openhouse-jtoyoda**
I am just using this to be able to add comments

### CODE REVIEW

File: builder_data_loader.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+                create_community(
+                    convert_community_data_to_json(
+                        row, 
+                        content_space, 
+                        openhouse_fields, 
+                        customer_fields, 
+                        entity_type),
+                    graphql_endpoint)
+        #floorplan    
+        if(entity_type == OPENHOUSE_ENTITYNAME_FLOORPLAN):
```
Comment (openhouse-jtoyoda) [inline builder_data_loader.py:45]: This should be elif

File: builder_data_loader.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+def get_csv_file_path(file_name):
+    return os.path.join(os.path.dirname(os.getcwd()),file_name)
+
+def main():
+    logger = get_logger('builder-data-loader')
+
+    # parse commandline arguments
+    args = parser.parse_args()
+    entity_type = args.entity
+    content_space = OPENHOUSE_STAGING_CONTENTSPACE if args.stg else args.contentspace
```
Comment (openhouse-jtoyoda) [inline builder_data_loader.py:87]: We have multiple staging spaces. May want to still just take in content space. Use the args.stg to set graphql though below

File: builder_data_loader.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+def load_entity_data(entity_type, content_space, csv_file_path, graphql_endpoint):
+
+    customer_fields = get_customer_field_names(
+        get_builder_config_path(content_space),
+        CONFIG_ENTITY_DATA_MAPPINGS[entity_type])
+
+    openhouse_fields = get_openhouse_field_names(
+        get_builder_config_path(content_space),
+        CONFIG_ENTITY_DATA_MAPPINGS[entity_type])
```
Comment (openhouse-jtoyoda) [inline builder_data_loader.py:29]: This very much feels like it should be a dict (map in other languages) data structure? Using identical indexes to map between them is very dangerous

File: data_loader_utils.py [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+
+logger = get_logger('builder_data_loader.data_loader_utils')
+
+def get_builder_config_path(content_space):
+    return os.path.join(os.getcwd(),'BuilderConfig\\config_{}.ini'.format(content_space))
+
+def generate_builder_id(display_name):
+    return display_name.strip().replace(" ", "_").upper()
+
+def convert_community_data_to_json(row, content_space, openhouse_fields, customer_fields, entity_type):
```
Comment (openhouse-jtoyoda) [inline data_loader_utils.py:18]: Isn't entity_type always going to be the same

File: data_loader_utils.py [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+def convert_floorplan_data_to_json(row, content_space, openhouse_fields, customer_fields, entity_type):
+
+    home_model = {}
+    builder_id=''
+    community_mapping = get_community_mappings(get_builder_config_path(content_space),'CommunityMappings')
+    community_pricing=[]
+
+    for field in customer_fields:
+        openhouse_field_name = openhouse_fields[customer_fields.index(field)]
+        if((openhouse_field_name == OPENHOUSE_FIELDNAME_HALFBATHS
```
Comment (openhouse-jtoyoda) [inline data_loader_utils.py:46]: An "in" statement might be more concise

File: BuilderConfig/config_de9z6jp6yx7r.ini [hunk 6]
```diff
@@ -0,0 +1,53 @@
+[CommunityDataMappings]
```
[the remark this answers is not in the document]
Comment (paulabhilash) [reply BuilderConfig/config_de9z6jp6yx7r.ini:1]: this is added

[the remark this answers is not in the document]
Comment (paulabhilash) [reply BuilderConfig/config_de9z6jp6yx7r.ini:1]: I would like to keep the header row manadatory for the script to work. I have now mentioned it in the README.

File: README.md [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+  
+  > pipenv run python builder_data_loader.py --entity inventory --contentspace de9z6jp6yx7r --file spechomes.csv --stg
+
+## CSV File Requirements
+To successfully import CSV files, the must have the header row as the first row. It should contain the names of the entity type properties corresponding to the data in each column.
+
+
+## Builder Config file
+The builder config file should be used to map the column names to the graphql fields.
+
```
Comment (paulabhilash) [inline README.md:64]: CSV file and builder config

File: README.md [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+    
+-  The community column names should be mapped to their respective community name/builder id under the 'CommunityMappings' section in the config.
+
+Example:
+
+        [CommunityMappings]
+        location1=BECKLEY
+        location2=LINCOLNTON
+        location3=GREENSBORO
+        location4=MARTINSVILLE
```
Comment (paulabhilash) [inline README.md:95]: On floorplan community pricing

File: data_loader_utils.py (hunk 5 above)
Comment (paulabhilash) [reply data_loader_utils.py]: good suggestion. I have moved them into a list of numeric data fields

File: data_loader_utils.py (hunk 4 above)
Comment (paulabhilash) [reply data_loader_utils.py]: removed it

File: builder_data_loader.py (hunk 3 above)
Comment (paulabhilash) [reply builder_data_loader.py]: moved away from using the index to get the filed names

File: builder_data_loader.py (hunk 2 above)
Comment (paulabhilash) [reply builder_data_loader.py]: updated

File: builder_data_loader.py (hunk 1 above)
Comment (paulabhilash) [reply builder_data_loader.py]: updated

**Review by paulabhilash — APPROVED** [review]

File: data_loader_utils.py [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+    input_data = dict(contentSpaceId=content_space)
+
+    data_mappings = get_data_mappings(get_builder_config_path(content_space),
+        CONFIG_ENTITY_DATA_MAPPINGS[OPENHOUSE_ENTITYNAME_COMMUNITY])
+    
+    for key in data_mappings:
+        openhouse_field_name = data_mappings[key]
+
+        if(openhouse_field_name == OPENHOUSE_FIELDNAME_DISPLAYNAME):
+            input_data["builderId"]=generate_builder_id(row[key])
```
Comment (openhouse-jtoyoda) [inline data_loader_utils.py:31]: "builderId" should be a constant

File: builder_data_loader.py [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+        elif(entity_type == OPENHOUSE_ENTITYNAME_FLOORPLAN):
+            #floorplan 
+            for row in csv_reader:
+                create_floorplan(
+                    convert_floorplan_data_to_json(
+                        row, 
+                        content_space),
+                    graphql_endpoint)   
+        elif(entity_type == OPENHOUSE_ENTITYNAME_INVENTORY):
+            #quickpossesions
```
Comment (openhouse-jtoyoda) [inline builder_data_loader.py:44]: typo

Comment (openhouse-jtoyoda) [reply builder_data_loader.py:44]: This comment is also unnecessary

File: data_loader_utils.py [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+
+import os
+import csv
+import json
+from openhouse_constants import *
+from load_builder_config import *
+from script_log_handler import *
+
+logger = get_logger('builder_data_loader.data_loader_utils')
+numeric_data_fields = [OPENHOUSE_FIELDNAME_HALFBATHS, OPENHOUSE_FIELDNAME_BATHS, OPENHOUSE_FIELDNAME_BEDS, OPENHOUSE_FIELDNAME_FLOORSIZE, OPENHOUSE_FIELDNAME_PRICE]
```
Comment (openhouse-jtoyoda) [inline data_loader_utils.py:11]: Much nicer

File: data_loader_utils.py [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+            community[openhouse_field_name] = row[key]
+
+    input_data[OPENHOUSE_ENTITYNAME_COMMUNITY]=community
+    
+    return json.dumps(dict(input=input_data))
+
+def convert_floorplan_data_to_json(row, content_space):
+
+    home_model = {}
+    community_mapping = get_community_mappings(get_builder_config_path(content_space),'CommunityMappings')
```
Comment (openhouse-jtoyoda) [inline data_loader_utils.py:43]: CommunityMappings should be a constant

File: data_loader_utils.py [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+            builder_id = generate_builder_id(row[key])
+        elif(key in community_mapping.keys() and row[key] != ''):
+            if(row["price_{}".format(key)] == ''):
+                community_pricing.append(dict(idType="BUILDER_ID",id=community_mapping[key],price=0))
+            else:
+                community_pricing.append(dict(idType="BUILDER_ID",id=community_mapping[key],price=row["price_{}".format(key)]))
+        elif(openhouse_field_name != ''):
+            home_model[openhouse_field_name] = row[key]
+
+    home_model["communityPricing"] = community_pricing
```
Comment (openhouse-jtoyoda) [inline data_loader_utils.py:65]: communityPricing here also should be an OPENHOUSE_FIELDNAME

**Review by openhouse-jtoyoda — COMMENTED** [review]
90% of the way there. Just a few more small things and should be good to go. Hope this helped

**Comment by paulabhilash** [comment]
> 90% of the way there. Just a few more small things and should be good to go. Hope this helped

Thanks for the comments! This was definitely helpful. 
I have updated them by making them constants and also extracted some of the string formattings to functions.

**Comment by openhouse-jtoyoda** [comment]
Looks good!

**Pull request merged**